Add basic structure and layout for homescreen

oliverguenther opened the pull request:
This PR is a collaboration WIP on the homescreen feature.
### TODOS
- [X] Decide what to show to an anonymous user. Anonymous users will be shown all blocks that are eligible to them. Users can also configure a separate welcome text block in the settings.

Relevant work package (internal): https://community.openproject.org/work_packages/19722

oliverguenther committed:
Homescreen functionality

This commit adds the homescreen functionality to OpenProject,
replacing the default welcome screen.

This adds support for:

* Blocks of content for various privilege levels (latest projects, users, news, etc.)
* Links to common functionality (administration, community)

Relevant work package:
https://community.openproject.org/work_packages/19722

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-19722-home-screen.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-19722-home-screen.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

oliverguenther committed:
Add controller and routing specs

oliverguenther committed:
Optional: show welcome text on the homescreen

This commit extends the settings to allow `welcome_text` to appear
on the homescreen and/or the projects page (previous behavior), or
neither.

NobodysNightmare commented:
Probably not part of the acceptance criteria, but after seeing the mockups my first thought was "wow, those blocks allow for a very responsive design".

I thought that, because you are free to have 1, 2, 3 or more columns. As far as I can tell we used that to enable a 2 column and a 3 column layout. I'd strongly suggest also allowing a single column for narrow screens.
At least if it is as easy to do as I would imagine.

NobodysNightmare commented:
I did not yet check the code, so I only know what the UI tells me.
Not sure what **_Latest projects**_ are. I would expect the most recent (visible) projects added to the instance?
But I can only see two rather old projects for my local instance. It also does not seems to be the latest projects where I am a member... So either there is a bug (I would have to discover that in code) or I don't understand which projects are shown to me.

NobodysNightmare commented:
Just as feedback, the icon besides "OpenProject" looks rather murky on default zoom:

![image](https://cloud.githubusercontent.com/assets/453584/9764144/a2bcc814-570d-11e5-84a4-d3ebc1085ded.png)

I also don't think this should read "**OpenProject**". Rather it should read "**MyCustomInstanceName**", shouldn't it?

NobodysNightmare commented:
### Link improvements

_Note: I merged comments regarding links into this comment._
- [x] **Report a bug**

Link is broken. It contains a superfluous `/` at the end, which prohibits correct rendering of that page.
- [x] **Professional support**

This might better link to https://www.openproject.com/enterprise-services/

Otherwise it will link to the place where you can create a new instance, which is especially surprising, when you are already on an op.com instance.
- [x] **Add and edit translations**

Probably better link here: https://crowdin.com/projects/opf

This is the homepage of our translations, promoting to translate everything, not only the core.
- [ ] **workflows**

This link is great! Actually it is so great, that I am asking myself, why the actual menu entry for workflows in the side menu is linking to `workflows/edit`, rather than `workflows` directly.
I think the one you've chosen here is the better one.
- [x] **advanced settings**

I am not sure how much I like the link text. "advanced" reads like **don't touch this**. However, many of the settings there are rather basic (e.g. "how do you want to name your instance?").
This is probably a deeper problem (the fact that we do not distinguish between advanced and basic anyway).

So probably the link text is good enough, just wanted to throw my cents in...
- [x] **On screen tips** (bottom row)

_Note: They were removed, I guess that means the problem is solved oO_

Only links to the FAQ, that was linked before... Not sure where you want the link to point to.

NobodysNightmare commented:
"Professional support" might better link to https://www.openproject.com/enterprise-services/

Otherwise it will link to the place where you can create a new instance, which is especially surprising, when you are already on an op.com instance.

NobodysNightmare commented:
> Adavanced Setttings

d-d-d-d-d-double ~~kill~~ typo:

Ad~~**a**~~vanced Sett~~**t**~~ings

oliverguenther commented:
> I thought that, because you are free to have 1, 2, 3 or more columns.

Homescreen doesn't put a constraint on that, we can actually let it slide into column mode for a smaller resolution. The minimal width currently is 1100px set for `html` and in this case, two columns are still fine.

> The "Report a bug" link is broken. It contains a superfluous / at the end

Wat? Oh..., angular.

> I also don't think this should read "OpenProject". Rather it should read "instance name", shouldn't it?

It does take the application title from the setting, and uses OpenProject only as a fallback

> Adavanced Setttings

Horsesome :horse:

NobodysNightmare commented:
> It does take the application title from the setting

Did not check after your rebase, but for me it didn't do that. The HTML title (the one shown as title of my browser tab) showed my changed instance name, while I could read "OpenProject" as the UI title of the homescreen.

oliverguenther commented:
You're right, I took the 'Software Name' setting instead of the app title, that's configurable in settings.

:ballot_box_with_check:

NobodysNightmare commented on app/assets/stylesheets/open_project_global/_variables.sass:
```diff
@@ -194,3 +194,9 @@ $user-avatar-mini-border-radius:                50px !default
 
 $select-element-padding:                        3px, 24px, 3px, 3px
+
+$homescreen-content-bg-color:                   white
```
`$homescreen-content-bg-color = $body-background`?

oliverguenther replied:
:ballot_box_with_check:

NobodysNightmare commented on app/assets/stylesheets/open_project_global/_variables.sass:
```diff
@@ -194,3 +194,9 @@ $user-avatar-mini-border-radius:                50px !default
 $select-element-padding:                        3px, 24px, 3px, 3px
+
+$homescreen-content-bg-color:                   white
+$homescreen-block-bg-color:                     white
```
I'd also use a variable here. I was looking out for something like `$block-color`, but could not find anything... so probably also `$body-background`...

oliverguenther replied:
:ballot_box_with_check:

NobodysNightmare commented on app/assets/stylesheets/open_project_global/_variables.sass:
```diff
@@ -194,3 +194,9 @@ $user-avatar-mini-border-radius:                50px !default
+$homescreen-block-bg-color:                     white
+$homescreen-footer-bg-color:                    #F8F8F8
+$homescreen-block-border-color:                 #EDEDED
+$homescreen-icon-color:                         #7B827B
```
Those are the homescreen footer icons right?

Asking because there are other icons on the homescreen, receiving the font color of `#333`

oliverguenther replied:
Renamed to `$homescreen-footer-icon-color` for clarity.

NobodysNightmare commented on app/assets/stylesheets/open_project_global/_variables.sass:
```diff
@@ -194,3 +194,9 @@ $user-avatar-mini-border-radius:                50px !default
+
+$homescreen-content-bg-color:                   white
+$homescreen-block-bg-color:                     white
+$homescreen-footer-bg-color:                    #F8F8F8
```
There are quite a few occurences of `#F8F8F8` in the global variables.

Not sure if it is a good idea to either assign one of the already existing ones here or to introduce a common color that most of those occurences use.

oliverguenther replied:
I'd suggest to make this `$gray-light` for the time being. I'd like to keep this separate color an option for themes.

NobodysNightmare commented on app/assets/stylesheets/open_project_global/_variables.sass:
```diff
@@ -194,3 +194,9 @@ $user-avatar-mini-border-radius:                50px !default
+$homescreen-content-bg-color:                   white
+$homescreen-block-bg-color:                     white
+$homescreen-footer-bg-color:                    #F8F8F8
+$homescreen-block-border-color:                 #EDEDED
```
We are not using this color anywhere else...

E.g. the boxes at `http://localhost:5000/projects/:id` have a border color of `#EAEAEA` (probably `$content-default-border-color`).

Here I would advise to use `$homescreen-block-border-color = $content-default-border-color`

oliverguenther replied:
I'm fine with that :ballot_box_with_check:

NobodysNightmare commented on app/assets/stylesheets/specific/homescreen.sass:
```diff
@@ -0,0 +1,144 @@
+          margin-bottom: 10px
+
+
+.homescreen-links
```
The naming of homescreen CSS classes is inconsistent.

Some read `homescreen-something` others read `homescreen--something`. According to the naming scheme we are using I'd say the component is the homescreen, so we want to use `homescreen--something`

edit: Probably there are exemptions... `homescreen-blocks` sounds legit, since it is a specific component used in the homescreen (and has its own `homescreen-blocks--something`)

edit2: looking further down, `homescreen-links` now also looks somewhat legit, at least it has own `--somethings`... naming things is hard :horse:

oliverguenther replied:
I'm also not sure on the exact hierarchy of these components. I think both `homescreen-links` and `homescreen--links` would be okay, IF there are more styles that relate to it..

I'm going to assume a `homescreen--` namespace and change all according to it.

NobodysNightmare commented on app/assets/stylesheets/specific/homescreen.sass:
```diff
@@ -0,0 +1,144 @@
+.homescreen-header
+  margin-left: 10px
+
+.homescreen-welcome
```
Outdated selector?

I can't find any element with that class...

oliverguenther replied:
True, has been replaced again. I'll remove it

NobodysNightmare commented:
Not sure how this happened, but there is some weird wrapping going on:

![ag...o](https://cloud.githubusercontent.com/assets/453584/9786781/09ab8e84-57bd-11e5-96c7-87076ccbd769.png)
1. this word should not be wrapped
2. there should be no wraps in the middle of a word

NobodysNightmare commented on app/assets/stylesheets/specific/homescreen.sass:
```diff
@@ -0,0 +1,144 @@
+        content: "\e03f"
+        display: inline-block
+        font-size: 0.6rem
+        margin-right: 10px
```
I felt that the distance between the "➔" and thelist text was rather large.
Checking the visual it also uses a smaller distance.

Is the distance between arrow and tex intended like that?

![image](https://cloud.githubusercontent.com/assets/453584/9787109/f04ee668-57bf-11e5-9c2e-94d076ef31b6.png)

If not: There is a marging **and** a padding on the right-hand side of the button. Probably only one of both is intended.

NobodysNightmare commented on config/locales/en.yml:
```diff
@@ -800,6 +821,7 @@ en:
   label_information_plural: "Information"
   label_integer: "Integer"
   label_internal: "Internal"
+  label_invite_user: "Invite user"
```
Why is that called "Invite user"?

It links to the create user form and that would be an apropriate label: "Create user".

I see that in the SaaS context there will be an invite feature, but that is nothing for on-premise OpenProject, is it?

oliverguenther replied:
No, invite user will hit the core as well, and does so in the next release. Niels asked me to name it this way.

NobodysNightmare commented on app/views/homescreen/index.html.erb:
```diff
@@ -0,0 +1,62 @@
+  </a>
+<% end %>
+</section>
+<% unless homescreen_socialmedia.empty? %>
```
`if homescreen_socialmedia.any?`

NobodysNightmare commented on app/helpers/homescreen_helper.rb:
```diff
@@ -0,0 +1,106 @@
+  # currently not present in the core
+  def homescreen_socialmedia
+    []
+  end
```
Just to understand. Any plugin (e.g. one of SaaS) can insert things here and then this will render additional links on the welcome screen?

I'd probably generalize this then. Do we need to know that those are "social" links? Why not "additional" links?
I'd provide an interface that anyone could use to inject links. E.g. for certain enterprises they could link prominent documents this way (privacy statements or whatever, internal portals or whatever).

Anyone can use this interface anyway, this is why I would use more generic wording here. It is probably only a naming decision after all...

NobodysNightmare commented on app/views/homescreen/blocks/_administration.html.erb:
```diff
@@ -0,0 +1,47 @@
+  <%= l(:label_administration) %>
+</h3>
+
+<ul class="homescreen--arrow-links">
```
I am asking myself whether those lists will occur in other places of the application. Then `-arrows` or `-arrow-items` might also be good class names...

oliverguenther replied:
I do not know of any interest in this style. I'd suggest that we make this a styleguide as soon as interest in arrow links spikes due to the amazing grace of the homescreen. (That is, when we have another use case)

NobodysNightmare commented on app/views/homescreen/blocks/_my_account.html.erb:
```diff
@@ -0,0 +1,22 @@
+  <li>
+    <%= link_to l('account.delete'), delete_my_account_info_path, title: l('account.delete') %>
+  </li>
+  <% end %>
```
I see that the content of this block might change once the "My Account" page was restructured. But do we want to have such a prominent "delete my account" link on the start page?

This is not like a common task you want to reach quickly ^^

oliverguenther replied:
Probably not, I'll reassure this one.

Edit: Removed :ballot_box_with_check:

NobodysNightmare commented on app/views/homescreen/index.html.erb:
```diff
@@ -0,0 +1,62 @@
+<% unless homescreen_socialmedia.empty? %>
+  <section class="homescreen-social-links">
+  <% homescreen_socialmedia.each do |link| %>
+    <a href="<%= link[:url] %>" title="<%= link[:name] %>" class="<%= link[:link_class] %>"></a>
```
In the spirit of my "generalize social media" comment:

I'd also add a `link[:caption]` that will be displayed as text below the Icon (I assume that `link_class` will add fancy icons).
We can leave it empty for our own social media purposes and it won't harm anyone ^^

oliverguenther committed:
Simple API for managing the homescreen content

This commit alleviates the homescreen management from both core and
plugins, in order to simply add and remove blocks.

Also contains refactoring from review
* Consistent block and variable usage
* Removed unused labels
* Avoid word breaks in multiline enumerations and remove margin from enumerations

NobodysNightmare commented on config/locales/en.yml:
```diff
@@ -753,6 +773,7 @@ en:
   label_duplicated_by: "duplicated by"
   label_duplicates: "duplicates"
   label_edit: "Edit"
+  label_email_notification: "Email notification"
```
unused...

NobodysNightmare commented on config/locales/en.yml:
```diff
@@ -916,7 +940,9 @@ en:
   label_principal_search: "Search for users or groups"
+  label_professional_support: "Professional support"
   label_profile: "Profile"
+  label_project_administration: "Project Administration"
```
unused

NobodysNightmare commented on config/locales/en.yml:
```diff
@@ -637,12 +637,28 @@ en:
   instructions_after_logout: "You can sign in again by clicking %{signin}."
   instructions_after_error: "You can try to sign in again by clicking %{signin}. If the error persists, ask your admin for help."
 
+  label_access_token: "Access token"
```
unused

NobodysNightmare commented on config/locales/en.yml:
```diff
@@ -637,12 +637,28 @@ en:
+      glossary: "Glossary"
+      shortcuts: "Shortcuts"
+      forums: "Forums"
+      onscreen_tips: "On screen tips"
```
you deleted that one.

unused...

NobodysNightmare commented on spec/controllers/homescreen_controller_spec.rb:
```diff
@@ -0,0 +1,120 @@
+        expect(response).to render_template(partial: 'homescreen/blocks/_news')
+      end
+
+      context 'with enabled welcome block' do
```
there is no assertion that the welcome block is missing otherwise (not contained in `all_blocks`)

oliverguenther replied:
:ballot_box_with_check:

oliverguenther commented:
> Not sure how this happened, but there is some weird wrapping going on:

This is fixed now, props to @HDinger 

> Just to understand. Any plugin (e.g. one of SaaS) can insert things here and then this will render additional links on the welcome screen?

I've extracted blocks and links into a Homescreen manager, which can be updated by plugins as they like, since they are effectively shallow wrappers on the item arrays.

Instead of the empty social icons content, I've allowed plugins to add custom content there with `call_hook`.

NobodysNightmare commented:
LGTM!

Waiting for travis and then :horse: :horse: :horse:

ulferts commented:
This introduced some inconsistencies as [noted](https://community.openproject.org/work_packages/19722#note-26):

Homepage

![image](https://cloud.githubusercontent.com/assets/617519/9808519/dc21e46a-585f-11e5-9b08-78f0634ca9a5.png)

vs MyPage

![image](https://cloud.githubusercontent.com/assets/617519/9808513/cee36936-585f-11e5-9790-2668834a147e.png)

vs Project Overview

![image](https://cloud.githubusercontent.com/assets/617519/9808525/f831e254-585f-11e5-9954-ed0cac08d810.png)

I guess the my_project_page plugin to be inconsistent as well.

ulferts commented:
I opened up a [WP](https://community.openproject.org/work_packages/21473).
